src/main/java/org/terasology/web/serverAdminManagement/AdminPermissionManager.java

```diff
@@ -0,0 +1,206 @@
+                if (permissions.getValue().get(permissionType)) {
+                    permissionManager.addPermission(clientInfo, consolePermissionsMap.get(permissionType));
+                } else {
+                    permissionManager.removePermission(clientInfo, consolePermissionsMap.get(permissionType));
+                }
+            }
+        }
+    }
+
+    public void giveAllPermissionsToAdmin(String adminId) {
+        setAdminPermissions(adminId, new Pair<>(adminId, generatePermissionMap(true)));
+    }
+
+    @SuppressWarnings({"SuspiciousToArrayCall", "SuspiciousMethodCalls"})
+    public void setAdminPermissions(String adminId, Pair<String, Map<PermissionType, Boolean>> newPermissions) {
+        Pair<String, Map<PermissionType, Boolean>> permission = getPermissionsOfAdmin(adminId);
+        serverAdminPermissions.remove(permission);
+        // Hack: somewhere along the line, the values of newPermissions get changed to Strings instead of PermissionTypes.
+        // This causes casting errors unless they are all turned back into PermissionTypes.
+        Map<PermissionType, Boolean> fixedNewPermissions = new HashMap<>();
+        for (String permissionType : newPermissions.getValue().keySet().toArray(new String[0])) {
+            fixedNewPermissions.put(PermissionType.valueOf(permissionType), newPermissions.getValue().get(permissionType));
+        }
+        serverAdminPermissions.add(new Pair<>(adminId, fixedNewPermissions));
+        try {
+            saveAdminPermissionList();
+        } catch (IOException e) {
+            logger.error("cannot save the admin permission list after setting a permission", e);
+        }
+        EntityRef playerToChange = EntityRef.NULL;
+        for (Client player : networkSystem.getPlayers()) {
+            if (player.getId().equals(adminId)) {
+                playerToChange = player.getEntity();
+            }
+        }
+        if (playerToChange != EntityRef.NULL) {
+            updateAdminConsolePermissions(adminId, playerToChange);
+        }
+    }
+
+    public void addAdmin(String id) {
+        serverAdminPermissions.add(new Pair<>(id, generatePermissionMap(false)));
+        onListChanged.run();
+    }
+
+    public void removeAdmin(String id) {
+        for (Pair<String, Map<PermissionType, Boolean>> adminPermission : serverAdminPermissions) {
+            if (adminPermission.getKey().equals(id)) {
+                serverAdminPermissions.remove(adminPermission);
+            }
+        }
+        Pair<String, Map<PermissionType, Boolean>> adminPermission = getPermissionsOfAdmin(id);
+        serverAdminPermissions.remove(adminPermission);
+        onListChanged.run();
+    }
+
+    public Pair<String, Map<PermissionType, Boolean>> getPermissionsOfAdmin(String id) {
+        for (Pair<String, Map<PermissionType, Boolean>> adminPermission : serverAdminPermissions) {
+            if (adminPermission.getKey().equals(id)) {
+                return adminPermission;
```

Comment:
In my understanding you could only return the right part (or value) of the pair, which has the actual information about permissions - the left part (or key) is the id, which the caller of this method already knows since it passes it as argument, correct?